expo/notifications.go

```diff
@@ -0,0 +1,198 @@
+package expo
+
+// Adapted from https://github.com/oliveroneill/exponent-server-sdk-golang.
+// We aren't using that package directly because it does not support the
+// use of Expo Access Tokens or handling the Receipt ID. There are already
+// PR's open for these issues, but it seems like the package is no longer
+// maintained.
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"os"
+)
+
+type PushMessage struct {
+	To    string `json:"to"`
+	Title string `json:"title,omitempty"`
+	Body  string `json:"body"`
+}
+
+type Response struct {
+	Data   []PushResponse      `json:"data"`
+	Errors []map[string]string `json:"errors"`
+}
+
+type PushResponse struct {
+	PushMessage PushMessage
+	ID          string            `json:"id"` // receipt
+	Status      string            `json:"status"`
+	Message     string            `json:"message"`
+	Details     map[string]string `json:"details"`
+}
+
+func (r *PushResponse) isSuccess() bool {
+	return r.Status == SuccessStatus
+}
+
+func (r *PushResponse) ValidateResponse() error {
+	if r.isSuccess() {
+		return nil
+	}
+	err := &PushResponseError{
+		Response: r,
+	}
+	// Handle specific errors if we have information
+	if r.Details != nil {
+		e := r.Details["error"]
+		if e == ErrorDeviceNotRegistered {
+			return &DeviceNotRegisteredError{
+				PushResponseError: *err,
+			}
+		} else if e == ErrorMessageTooBig {
+			return &MessageTooBigError{
+				PushResponseError: *err,
+			}
+		} else if e == ErrorMessageRateExceeded {
+			return &MessageRateExceededError{
+				PushResponseError: *err,
+			}
+		}
+	}
+	return err
+}
+
+// SuccessStatus is the status returned from Expo on a success
+const SuccessStatus = "ok"
+
+// ErrorDeviceNotRegistered indicates the token is invalid
+const ErrorDeviceNotRegistered = "DeviceNotRegistered"
+
+// ErrorMessageTooBig indicates the message went over payload size of 4096 bytes
+const ErrorMessageTooBig = "MessageTooBig"
+
+// ErrorMessageRateExceeded indicates messages have been sent too frequently
+const ErrorMessageRateExceeded = "MessageRateExceeded"
+
+type PushResponseError struct {
+	Response *PushResponse
+}
+
+func (e *PushResponseError) Error() string {
+	if e.Response != nil {
+		return e.Response.Message
+	}
+	return "Unknown push response error"
```

Comment:
nit: error message strings conventionally start with a lower-case letter

Reply:
I'm using the official expo package now for this functionality.